[GOBBLIN-2168] Compare with the current time in UTC

linweihs opened the pull request:
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [GOBBLIN-2168](https://issues.apache.org/jira/browse/GOBBLIN-2168) 

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):

In existing trash cleaner, the comparison between snapshotTime and the current time is NOT done in the same time zone, this leads to longer hard deletion time due to time zone difference.

the fix ensure that the comparison between snapshotTime and the current time is done in the same time zone



### Tests
- [x] My PR adds the following unit tests :

Implement a unit test class `[TimeBasedSnapshotCleanupPolicyTest.java](https://github.com/apache/gobblin/compare/master...linweihs:incubator-gobblin:welin/trash?expand=1#diff-59e30e386f99115a2511854af4f01812ea1cf832430c696bb48182af91713f77)` ensuring the comparison are asserted expectedly.


### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

Will-Lo commented on gobblin-data-management/src/main/java/org/apache/gobblin/data/management/trash/TimeBasedSnapshotCleanupPolicy.java:
```diff
@@ -41,6 +42,7 @@ public TimeBasedSnapshotCleanupPolicy(Properties props) {
     DateTime snapshotTime = Trash.TRASH_SNAPSHOT_NAME_FORMATTER.parseDateTime(snapshot.getPath().getName());
-    return snapshotTime.plusMinutes(this.retentionMinutes).isBeforeNow();
+    // To ensure that the comparison between snapshotTime and the current time is done in the same time zone
+    return snapshotTime.plusMinutes(this.retentionMinutes).isBefore(DateTime.now(DateTimeZone.UTC));
```
Should there be a configuration to ensure that it is using UTC compared to the old behavior? You don't want existing retention jobs to have unintended consequences due to change in behavior that is not explicit

Will-Lo replied:
I think it would be more extensible if we had something like the following:
```
public static final String RETENTION_SNAPSHOT_TIMEZONE = "gobblin.trash.snapshot.retention.timezone"
Timezone retentionSnapshotTimezone;
...
this.retentionSnapshotTImezone = props.containsKey(RETENTION_SNAPSHOT_TIMEZONE) ? Timezone.getTimeZone(props.getProperty(RETENTION_SNAPSHOT_TIMEZONE)) : DateTimeZone.getDefault();
...
return snapshotTime.plusMinutes(this.retentionMinutes).isBefore(DateTime.now(this.retentionSnapshotTimezone));
```
This allows the retention snapshot be more configurable in the long term and more explciit what the behavior is.

linweihs replied:
thanks for the suggestion, this is indeed much cleaner and more reasonable

linweihs commented:
updated the implementation according to comment suggested. thanks for the review

linweihs commented on gobblin-data-management/src/test/java/org/apache/gobblin/data/management/trash/TimeBasedSnapshotCleanupPolicyTest.java:
```diff
@@ -0,0 +1,98 @@
+                Integer.toString(SNAPSHOT_RETENTION_POLICY_MINUTES_DEFAULT)));
+            this.MOCK_CURRENT_TIME = mockCurrentTime;
+            this.retentionSnapshotTimezone = props.containsKey(RETENTION_SNAPSHOT_TIMEZONE) ? DateTimeZone.forID(props.getProperty(RETENTION_SNAPSHOT_TIMEZONE)) : DateTimeZone.getDefault();
+        }
```
this is helpful ! i've addressed according to the suggestion, thanks for the reviewing :)

phet commented on gobblin-data-management/src/test/java/org/apache/gobblin/data/management/trash/TimeBasedSnapshotCleanupPolicyTest.java:
```diff
@@ -0,0 +1,90 @@
+            this.MOCK_CURRENT_TIME = mockCurrentTime;
+        }
+
+        @Override
+        public boolean shouldDeleteSnapshot(FileStatus snapshot, Trash trash) {
+            DateTime snapshotTime = Trash.TRASH_SNAPSHOT_NAME_FORMATTER.parseDateTime(snapshot.getPath().getName());
+            return snapshotTime.plusMinutes(this.retentionMinutes).isBefore(this.MOCK_CURRENT_TIME.withZoneRetainFields(this.retentionSnapshotTimezone));
+        }
```
as *THE ONLY* method of the class-under-test, the test itself can in no way meaningfully verify the impl, when this mock actually reimplements that method!

I see your challenge: you'd like to hard-code a known date-time in the test, but that's not possible because the `TimeBasedSnapshotCleanupPolicy` isn't parameterized by a source/provider of the current date-time.

typically [that is the canonical testing pattern](https://stackoverflow.com/a/9410093) I would recommend, but in this particular case, where the crux of the impl comes down to the actual use of `DateTime.now(DateTimeZone)`, I'd suggest instead NOT to hard-code a known time but rather to calculate one based on the true current time when the test runs.

e.g. set retention at 600 mins (10 hours) and render a path that is between 12 and 13 hours ago in UTC.  then instantiate an instance using TZ.UTC to verify `Assert.assertTrue(x.shouldDeleteSnapshot(...))` and instantiate another policy using TZ.PST to verify `assertFalse`.  (there's nothing special about PST, but just choose a TZ that is at least >= 4 hours later than UTC.)

then perhaps round out the test by then generating a path that is > 10 hours, even in PST, and also another that's < 9 hours ago in UTC

linweihs replied:
thanks, i've updated my newest findings and test accordingly

phet commented on gobblin-data-management/src/test/java/org/apache/gobblin/data/management/trash/TimeBasedSnapshotCleanupPolicyTest.java:
```diff
@@ -0,0 +1,90 @@
+        Trash trash = trashTestBase.trash;
+
+        // Create dummy paths
+        FileStatus fs1 = new FileStatus(0, true, 0, 0, 0, 
+                                        new Path(trash.getTrashLocation(), new DateTime(2024, 10, 29, 1, 0, DateTimeZone.UTC).toString(Trash.TRASH_SNAPSHOT_NAME_FORMATTER)));
```
speaking of mocks, it seems reasonable to mock the `FileStatus` to only return `.getPath()`.  that demonstrates the check is based solely on the path itself and not e.g. on the modtime

linweihs replied:
updated. Mock the FileStatus object to return the path

phet reviewed (changes requested):
good impl fix here... but the test needs reworking

linweihs commented:
**Updated 2024/11/20:**
I found out a bug internally where theres discrepancy between java.utils.Properties and Azkaban prop object during props passing, thats results in parent config is not pass all the way down to `TimeBasedSnapshotCleanupPolicy.java` in runtime.

I've reverted TimeBasedSnapshotCleanupPolicy to original code, and tested are expected.  the However, i've added unit tests `TimeBasedSnapshotCleanupPolicyTest` ensuring the test coverage.

During investigation, also found a bug (in Trash.java) where Trash use `props.contains(TRASH_CLASS_KEY))` which leads to unexpected evalation whereas the intention is to `props.containsKey(TRASH_CLASS_KEY))`
![Screenshot 2024-11-20 at 2 48 37 PM](https://github.com/user-attachments/assets/c4bb17d5-8928-4f99-8707-0694f76a209f)
